### PR TITLE
Double the Donation settings have been reviewed

### LINKED ISSUE #2: Double the Donation settings should be reviewed
## Details

The Double the Donation settings page should follow the standards set by existing GiveWP plugins so that it is more discoverable and maintainable.

## Visuals

image

1. The `Double the Donation` tab is nested under `General > Settings`, but other add-ons have top-level tabs. It's unclear to me whether this is intentional or not.
2. This registration form link should probably use a short URL in case we need to update it in the future.
3. This docs URL does not exist

## Acceptance Criteria

-   [x] The settings tab position is reviewed and moved if necessary.
-   [x] The registration form link uses a short URL.
-   [x] The docs URL in the bottom right corner of the settings page is fixed.

### PR DESCRIPTION
**Pull request by lindseymacmillan**
Resolves #2 

## Description
This PR addresses a handful of issues related to admin settings for the Double the Donation add-on. Specifically, the PR moves the settings into their own top-level tab, and updates form links.

## Affects
This PR affects the double the donation ServiceProvider (where setting pages are loaded) and replaces the SettingsTab class with the SettingsPage class to handle rendering the settings page content.

## Visuals
n/a

## Pre-review Checklist
-   [ ] Acceptance criteria satisfied and marked in related issue
-   [x] Relevant `@since` tags included in DocBlocks
-   [x] Changes logged to the `Unreleased` section of `CHANGELOG.md`
-   [x] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

## Testing Instructions
1. Check that the Double the Donation settings are in their own tab of the Settings page
2. Check that the registration form link uses a short url
3. Check that the docs URL exists

### CODE REVIEW

File: src/DoubleTheDonation/AddonServiceProvider.php [hunk 1]
```diff
@@ -59,13 +62,13 @@ private function loadBackend() {
...
+		SettingsPage::registerPage( DoubleTheDonationSettingsPage::class );
 
-		// Will display html of the import donation.
-		Hooks::addAction('give_admin_field_dtd_intro',SettingsTab::class, 'renderIntro');
+		//Hooks::addFilter('plugin_action_links_' . GIVE_DTD_BASENAME, DoubleTheDonationSettingsPage::class, 'add_settings_link' );
 
-		// Add settings tab.
-		Give(SettingsTab::class)->addTab();
+		// Will display html of the import donation.
+		//Hooks::addAction('give_admin_field_dtd_intro',DoubleTheDonationSettingsPage::class, 'render_intro');
```
Comment (lindseymacmillan) [inline src/DoubleTheDonation/AddonServiceProvider.php]: @JasonTheAdams @kjohnson One road-block I am running into is displaying intro text and action links on the settings page. For some reason, when loaded via hooks which invoke functions off the SettingsPage class (in the same manor they were doing previously), I hit this error. It suggests to me that maybe the hooks are being fired before Give_Settings_Page is accessible? 

The above commented code is causing the error (when uncommented). 

Screen Shot 2020-10-30 at 1 04 19 PM

Comment (kjohnson) [reply src/DoubleTheDonation/AddonServiceProvider.php]: @henryholtgeerts  at first glance, I would think it is a loading order issue, but the hook should fire after GiveWP is loaded.

It looks like `abstract-admin-settings-page.php` is included dynamically, so the contexts in which the Give_Settings_Page class is available is limited.

I would suggest extracting `add_settings_link()` and `render_intro()` out of `src\DTD\SettingsPage`.

You can move `render_intro()` into a template file and add the filter for `add_settings_link()` inline of the `ServiceProvider::boot`.

Comment (lindseymacmillan) [reply src/DoubleTheDonation/AddonServiceProvider.php]: Ah - this makes sense! Thanks for taking a look. 🙌

**Comment by lindseymacmillan** [comment]
@mathetos This PR is ready short of your input on a docs URL and link for the associated blog post or readme file (instead of linking to docs).

File: src/DoubleTheDonation/AddonServiceProvider.php [hunk 2]
```diff
@@ -65,10 +65,10 @@ private function loadBackend() {
 		// Register settings page
 		SettingsPage::registerPage( DoubleTheDonationSettingsPage::class );
 
-		//Hooks::addFilter('plugin_action_links_' . GIVE_DTD_BASENAME, DoubleTheDonationSettingsPage::class, 'add_settings_link' );
+		Hooks::addFilter('plugin_action_links_' . GIVE_DTD_BASENAME, AddonServiceProvider::class, 'addSettingsLink' );
 
 		// Will display html of the import donation.
-		//Hooks::addAction('give_admin_field_dtd_intro',DoubleTheDonationSettingsPage::class, 'render_intro');
+		Hooks::addAction('give_admin_field_dtd_intro',AddonServiceProvider::class, 'renderIntro');
```
Comment (kjohnson) [inline src/DoubleTheDonation/AddonServiceProvider.php]: @henryholtgeerts `Hooks::add*` delays class instantiation until the hook is called, so it isn't needed when hooking a method from within that class.

**Review by kjohnson — CHANGES_REQUESTED** [review]

File: src/DoubleTheDonation/AddonServiceProvider.php (hunk 2 above)
Comment (lindseymacmillan) [reply src/DoubleTheDonation/AddonServiceProvider.php]: Done, ref: https://github.com/impress-org/give-double-the-donation/pull/6/commits/91ebea18992d36c09ee6644bc91636de3f12ca09

**Comment by JasonTheAdams** [comment]
@henryholtgeerts Apparently you and I were tweaking things at the same time. Hahah! I did clean up a few things. Take a look at 5e3d648 to see what changes I made.

**Review by JasonTheAdams — APPROVED** [review]
Looking good! 👍

**Pull request merged**